packages/features/ee/organizations/lib/OrganizationPermissionService.ts

```diff
@@ -0,0 +1,122 @@
+import { ORGANIZATION_SELF_SERVE_MIN_SEATS, ORGANIZATION_SELF_SERVE_PRICE } from "@calcom/lib/constants";
+import { prisma } from "@calcom/prisma";
+import type { TrpcSessionUser } from "@calcom/trpc/server/trpc";
+
+import { TRPCError } from "@trpc/server";
+
+type SeatsPrice = {
+  seats?: number | null;
+  pricePerSeat?: number | null;
+};
+
+export interface validatePermissionsIOrganizationPermissionService {
+  hasPermissionToCreateForEmail(targetEmail: string): Promise<boolean>;
+  hasPendingOrganizations(email: string, slug?: string): Promise<boolean>;
+  hasPermissionToModifyDefaultPayment(): boolean;
+  hasPermissionToMigrateTeams(teamIds: number[]): Promise<boolean>;
+  hasModifiedDefaultPayment(
+    input: {
+      billingPeriod?: string;
+    } & SeatsPrice
+  ): boolean;
+}
+
+export class OrganizationPermissionService {
+  constructor(private readonly user: NonNullable<TrpcSessionUser>) {}
+
+  async hasPermissionToCreateForEmail(targetEmail: string): Promise<boolean> {
+    return this.user.email === targetEmail || this.user.role === "ADMIN";
+  }
+
+  async hasCompletedOnboarding(email: string): Promise<boolean> {
+    const orgOnboarding = await prisma.organizationOnboarding.findUnique({
+      where: {
+        orgOwnerEmail: email,
+      },
+    });
+
+    return !!(orgOnboarding && orgOnboarding.isComplete);
+  }
+
+  hasPermissionToModifyDefaultPayment(): boolean {
+    return this.user.role === "ADMIN";
+  }
+
+  hasModifiedDefaultPayment(input: SeatsPrice & { billingPeriod?: string }): boolean {
+    return (
+      (input.billingPeriod !== undefined &&
+        input.billingPeriod !== null &&
+        input.billingPeriod !== "MONTHLY") ||
+      (input.seats !== undefined &&
+        input.seats !== null &&
+        input.seats !== ORGANIZATION_SELF_SERVE_MIN_SEATS) ||
+      (input.pricePerSeat !== undefined &&
+        input.pricePerSeat !== null &&
+        input.pricePerSeat !== ORGANIZATION_SELF_SERVE_PRICE)
+    );
+  }
+
+  async hasPermissionToMigrateTeams(teamIds: number[]): Promise<boolean> {
+    if (teamIds.length === 0) return true;
+
+    const teamMemberships = await prisma.membership.findMany({
+      where: {
+        userId: this.user.id,
+        team: {
+          id: {
+            in: teamIds,
+          },
+        },
+        role: {
+          in: ["OWNER", "ADMIN"],
+        },
+      },
+    });
+
+    return teamMemberships.length === teamIds.length;
+  }
+
+  async validatePermissions(
+    input: {
+      orgOwnerEmail: string;
+      teams?: { id: number; isBeingMigrated: boolean }[];
+      billingPeriod?: string;
+      slug: string;
+    } & SeatsPrice
+  ): Promise<boolean> {
+    if (!(await this.hasPermissionToCreateForEmail(input.orgOwnerEmail))) {
+      throw new TRPCError({
+        code: "UNAUTHORIZED",
+        message: "You do not have permission to create an organization for this email",
+      });
```

Comment:
NIT should we throw something other than a TRPC error here and also in other places in this class - so we don't fully scope this service to be used with TRPC in mind

Reply:
Yeah certainly that is the best way. Infact I plan to extend Error class and have specific errors where needed.